pkg/storage/redis/redis.go

```diff
@@ -208,6 +210,50 @@ func (db *DB) ClearDeleted(_ context.Context, cutoff time.Time) {
 	}
 }
 
+// Sync returns a channel to the caller, when there is a change to the version set,
+// sending message to the channel to notify the caller.
+func (db *DB) Sync(ctx context.Context) chan struct{} {
+	ch := make(chan struct{})
+	go func() {
+		c := db.pool.Get()
+		defer func() {
+			c.Close()
+			close(ch)
+		}()
+
+		// Setup notifications, we only care about changes to db.version_set.
+		if _, err := c.Do("CONFIG", "SET", "notify-keyspace-events", "Kz"); err != nil {
+			log.Error().Err(err).Msg("failed to setup redis notification")
+			return
+		}
+
+		psc := redis.PubSubConn{Conn: c}
+		if err := psc.PSubscribe("__keyspace*__:" + db.versionSet); err != nil {
+			log.Error().Err(err).Msg("failed to subscribe to version set channel")
+			return
+		}
+		for {
+			select {
+			case <-ctx.Done():
+				log.Error().Err(ctx.Err()).Msg("stop subscribing to version set channel")
+				return
+			default:
+			}
+			switch v := psc.Receive().(type) {
```

Comment:
We need to detect `ctx.Done` and terminate this call somehow. If I remember right this redis library has very poor context support. Some ideas:

1. Listen for ctx.Done() in a goroutine and then close the underlying connection, which should hopefully terminate the Receive() call
2. Use a relatively short timeout and retry on timeout errors. For example we could use a 1 second timeout.

Reply:
Fixed.